### DISCUSSION TITLE
Group::find('my-dn')->members()->recursive() returns groups, too?

### DISCUSSION THREAD

**Discussion by nilskretschmer** [DirectoryTree/LdapRecord#674] [Q&A]
Hello everyone,

we notice an unexpected behavior in this otherwise great library:

## Setup
We have an Active Directory with different User Groups. One User Group contains - let's say 5 Users. This group also contains a subgroup. The Subgroup also contains 3 users. Now we want to get all Users (members) of the base group and the subgroup. For this we use the `recursive()`-function on the `LdapRecord\Models\ActiveDirectory\Group`-Model:

```
use LdapRecord\Models\ActiveDirectory\Group as LdapGroup;

$ldapUsers = LdapGroup::find(config('ldap.app.users.group'))->members()->recursive()->get();
```

This `config('ldap.app.users.group')` returns the DN of our base group.

## The Problem
The function call above returns all users of the base group and the subgroup correctly. But the query also returns the subgroup as an object itself. This leads to the unexpected problem that you assume to get users only, but instead, you retrieve subgroups, too.

Is this indended or is there something wrong with our query?

Thank you in advance!

**Comment by stevebauman** [comment]
Hi @nilskretschmer!

This is intended behaviour, as the `members` relationship indicates anything that is a member of the group, including other groups, computers, contacts, and users:

https://github.com/DirectoryTree/LdapRecord/blob/db842db1e9a83269da8afbbf89fb447606503381/src/Models/ActiveDirectory/Group.php#L30-L37

If you'd only like users, I'd suggest either extending the built in `Group` model with your own and adding a `users` relationship:

```php
use LdapRecord\Models\Relations\HasMany;
use LdapRecord\Models\ActiveDirectory\User;
use LdapRecord\Models\ActiveDirectory\Group as BaseGroup;

class Group extends BaseGroup
{
    /**
     * The users relationship.
     */
    public function users(): HasMany
    {
        return $this->hasMany(User::class, 'memberof')->using($this, 'member');
    }
}
```

Or; Filtering the results to only include `User` models:

```php
use LdapRecord\Models\ActiveDirectory\User as LdapUser;
use LdapRecord\Models\ActiveDirectory\Group as LdapGroup;

$ldapUsers = LdapGroup::find(config('ldap.app.users.group'))->members()->recursive()->get()->filter(
    fn (Model $model) => $model instanceof LdapUser
);
```

Hope this helps!
